### PR TITLE
Adding Comments in BidirectionalSearch.java

### PR DESCRIPTION
**Pull request by kanishk1225**
@Manthan-R-Sheth please review this.

### CODE REVIEW

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java [hunk 1]
```diff
@@ -38,22 +38,37 @@
     @Override
 
     public BidirectionalActions<A> apply(Problem<A, S> originalProblem, Problem<A, S> reverseProblem) {
-        Node<A, S> node = newRootNode(originalProblem.initialState(), 0);
+        Node<A, S> node = newRootNode(originalProblem.initialState(), 0); //node -> starting point 
+        //if starting node is the goal node then return bidirectionalAction
```
Comment (globalworming) [inline core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:42]: please avoid obsolete comments http://google.com/#q=clean+code+comments
use comments where the code isn't self-explanatory

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java [hunk 2]
```diff
@@ -38,22 +38,37 @@
     @Override
 
     public BidirectionalActions<A> apply(Problem<A, S> originalProblem, Problem<A, S> reverseProblem) {
-        Node<A, S> node = newRootNode(originalProblem.initialState(), 0);
+        Node<A, S> node = newRootNode(originalProblem.initialState(), 0); //node -> starting point 
+        //if starting node is the goal node then return bidirectionalAction 
+        //which is the solution of the problem (total cost -> zero)
+        //First terminating condition
```
Comment (globalworming) [inline core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:44]: obsolete

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java [hunk 3]
```diff
@@ -38,22 +38,37 @@
...
         if (originalProblem.isGoalState(node.state())) {
             this.previousMeetingOfTwoFrontiers = node;
             this.nextNodeToBeEvaluated = null;
             bidirectionalActions = new BasicBidirectionalActions<>(fromInitialStatePart(), fromGoalStatePart());
             return bidirectionalActions;
         }
-        Node<A, S> revNode = newRootNode(reverseProblem.initialState(), 0);
+        
+        Node<A, S> revNode = newRootNode(reverseProblem.initialState(), 0);//revNode -> goal node
+        //front -> priority queue which has node to be explored when traversing from starting node
```
Comment (globalworming) [inline core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:53]: don't include variable names in comments. when names change, your comment is obsolete

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java [hunk 4]
```diff
@@ -38,22 +38,37 @@
...
         Queue<Node<A, S>> front = newFIFOQueue(node);
+        //back -> priority queue which has node to be explored when traversing from goal node
         Queue<Node<A, S>> back = newFIFOQueue(revNode);
+        //exploredFront -> map which has explored nodes marked when traversing from starting node 
         Map<S, Node<A, S>> exploredFront = newExploredMap(node.state());
+        //exploredBack -> map which has explored nodes marked when traversing from goal node
         Map<S, Node<A, S>> exploredBack = newExploredMap(revNode.state());
 
+        //while both the priority queue (i.e. front and back) are filled then that implies
```
Comment (globalworming) [inline core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:62]: wrong. they are currently filled with nodes, still there is no solution yet

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java [hunk 5]
```diff
@@ -38,22 +38,37 @@
...
+        //there is a path from both the end hence we traverse from both the end till we meet in the middle
+        //otherwise there is no solution
+        
+        //a loop is run till solution is found or
```
Comment (globalworming) [inline core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:66]: obsolete

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java [hunk 6]
```diff
@@ -63,45 +78,80 @@
...
+    	//exploredBack -> map which has explored nodes marked when traversing from goal node
+    	
+    	//problem contains information about the 
+    	//initial state, action that is to be performed, goal test, path cost and result function
+    	//which gives the result of the action 
+    	
+    	//if queue is not empty then there is a node to be explored
+    	if (!queue.isEmpty()) {
+    	    //exploring the first node in the queue
+    	    //next -> first node from queue
```
Comment (globalworming) [inline core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:101]: obsolete

File: core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java [hunk 7]
```diff
@@ -63,45 +78,80 @@
...
             }
         }
+    	//if queue is empty return null (i.e. failure)
         this.previousMeetingOfTwoFrontiers = null;
         return null;
     }
 
     private boolean isSolution(Node<A, S> solutionNode, boolean fromFront) {
+    	//solutionNode is a node which is checked whether it is a meeting of two frontiers or not
+    	//if solutionNode is not null
```
Comment (globalworming) [inline core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java:144]: obsolete

**Review by globalworming — CHANGES_REQUESTED** [review]
this is not in any way "clean code"
most comments are obsolet/redundant

**Comment by globalworming** [comment]
please don't merge @ctjoreilly 
this isn't helpful in any way

**Comment by globalworming** [comment]
> @Manthan-R-Sheth please review this.

if your colleague should review this, why do you open a pull request in the official repo? make a pull request to https://github.com/Manthan-R-Sheth/aima-java and not in the official repo

**Comment by kanishk1225** [comment]
@globalworming thanks for your suggestions. 
Actually, I have taken reference from the previously commented codes
https://github.com/aimacode/aima-java/blob/AIMA4e/core/src/main/java/aima/core/search/basic/informed/AStarSearch.java
Then also I will improve my commenting style.
Thanks once again.
If you have more suggestion you are most welcome.

**Comment by Manthan-R-Sheth** [comment]
@kanishk1225  @globalworming 
Comments can be added once the algorithm is released for the fourth edition as stated [here](https://github.com/aimacode/aima-pseudocode/issues/41). Comments will be written according to the released algorithm, just like how other implemented algorithms are commented. 
@ctjoreilly  can explain better.

**Comment by ctjoreilly** [comment]
Hi @kanishk1225 I am not going to merge this pull request as in general what @globalworming highlighted is correct in terms of the comments being too excessive. The example you referenced:

https://github.com/aimacode/aima-java/blob/AIMA4e/core/src/main/java/aima/core/search/basic/informed/AStarSearch.java

Is doing something slightly different. Yes, the comments can also be seen as excessive but the rationale is that the comments are the actual psuedocode from the book and even if there is a 1-1 mapping with the Java code the pseudocode comments are placed consistently above the corresponding java code to help someone who has read the book better understand the default/basic implementations of the algorithm.

**Pull request closed without merging**